src/dx/presence/mod.rs

```diff
@@ -0,0 +1,547 @@
+            ..Default::default()
+        }
+    }
+
+    /// Create a heartbeat request builder.
+    ///
+    /// This method is used to update state associated with `user_id` on
+    /// channels using `heartbeat` operation endpoint.
+    ///
+    /// Instance of [`HeartbeatRequestsBuilder`] returned.
+    pub fn set_state_with_heartbeat<U>(&self, state: U) -> HeartbeatRequestBuilder<T, D>
+    where
+        U: Serialize + Send + Sync + 'static,
+    {
+        self.heartbeat().state(state)
+    }
+}
+
+impl<T, D> PubNubClientInstance<T, D>
+where
+    T: Transport + Send + 'static,
+    D: Deserializer + 'static,
+{
+    /// Announce `join` for `user_id` on provided channels and groups.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
+    pub(crate) fn announce_join(
+        &self,
+        channels: Option<Vec<String>>,
+        channel_groups: Option<Vec<String>>,
+    ) {
+        self.configure_presence();
+
+        {
+            let slot = self.presence.read();
+            if let Some(presence) = slot.as_ref() {
+                presence.announce_join(channels, channel_groups);
+            }
+        };
+    }
+
+    /// Announce `leave` for `user_id` on provided channels and groups.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
+    pub(crate) fn announce_left(
+        &self,
+        channels: Option<Vec<String>>,
+        channel_groups: Option<Vec<String>>,
+    ) {
+        self.configure_presence();
+
+        {
+            let slot = self.presence.read();
+            if let Some(presence) = slot.as_ref() {
+                presence.announce_left(channels, channel_groups);
+            }
+        };
+    }
+
+    /// Complete presence configuration.
+    ///
+    /// Presence configuration used only with presence event engine.
+    #[cfg(feature = "std")]
+    #[allow(dead_code)]
+    pub(crate) fn configure_presence(&self) -> Arc<RwLock<Option<PresenceManager>>> {
+        {
+            let mut slot = self.presence.write();
+            if slot.is_none() {
+                *slot = Some(PresenceManager::new(self.presence_event_engine(), None));
+            }
+        }
+
+        self.presence.clone()
+    }
+
+    /// Presence event engine.
+    ///
+    /// Prepare presence event engine instance which will be used for `user_id`
+    /// presence announcement and management.
+    #[cfg(feature = "std")]
+    pub(crate) fn presence_event_engine(&self) -> Arc<PresenceEventEngine> {
+        let channel_bound = 3;
+        let (cancel_tx, cancel_rx) = async_channel::bounded::<String>(channel_bound);
+        let delayed_heartbeat_cancel_rx = cancel_rx.clone();
+        let wait_cancel_rx = cancel_rx.clone();
+        let runtime = self.runtime.clone();
+        let delayed_heartbeat_call_client = self.clone();
+        let heartbeat_call_client = self.clone();
+        let leave_call_client = self.clone();
+        let wait_call_client = self.clone();
```

Comment:
That's a big amount of pointers...
Do we need all of them ;o

I know that they're quite small in the memory, but question is if we need such a split ;o

Reply:
There are many of them, but we need to give closures to the event engine and use clone to make sure they last a long time. Is there a better way to do this?

Reply:
Okay. Closures explains me everything. 

I like closure but I think that we overuse them ;d. 
But let's keep it as it is. I was just asking why there are so many of them ;d